stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1865,239 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
+//   }
+//   return (result, pullback)
+// }
+// ```
+//
+extension Array : Differentiable where Element : Differentiable {
+  public var allDifferentiableVariables: AllDifferentiableVariables {
+    get {
+      return AllDifferentiableVariables(
+        elements: self.map { $0.allDifferentiableVariables })
+    }
+    set {
+      for i in indices {
+        self[i].allDifferentiableVariables = newValue.elements[i]
+      }
+    }
+  }
+
+  public struct AllDifferentiableVariables : Differentiable {
+    public typealias AllDifferentiableVariables = 
+      Array.AllDifferentiableVariables
+    public typealias TangentVector = Array.TangentVector
+    public typealias CotangentVector = Array.CotangentVector
+
+    public var elements: [Element.AllDifferentiableVariables]
+
+    public init(elements: [Element.AllDifferentiableVariables]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public func moved(
+      along direction: Array.TangentVector
+    ) -> Array.AllDifferentiableVariables {
+      precondition(self.elements.count >= direction.elements.count)
+      let elements = self.elements.indices.map {
+        self.elements[$0].moved(
+          along: $0 < direction.elements.count ? direction.elements[$0] : .zero)
+      }
+      return Array.AllDifferentiableVariables(elements: elements)
+    }
+
+    @inlinable
+    public func tangentVector(
+      from cotangent: Array.CotangentVector
+    ) -> Array.TangentVector {
+      precondition(self.elements.count >= cotangent.elements.count)
+      return Array.TangentVector(elements: self.elements.indices.map {
+        self.elements[$0].tangentVector(
+          from: $0 < cotangent.elements.count ? cotangent.elements[$0] : .zero)
+      })
+    }
+  }
+
+  public struct TangentVector : AdditiveArithmetic & Differentiable {
+    public typealias AllDifferentiableVariables = Array.TangentVector
+    public typealias TangentVector = Array.TangentVector
+    public typealias CotangentVector = Array.CotangentVector
+
+    public var elements: [Element.TangentVector]
+
+    public static var zero: Array.TangentVector {
+      return TangentVector(elements: [])
+    }
+    
+    public init(elements: [Element.TangentVector]) {
+      self.elements = elements
+    }
+
+    @inlinable
+    public static func + (
+      lhs: Array.TangentVector,
+      rhs: Array.TangentVector
+    ) -> Array.TangentVector {
+      return Array.TangentVector(
+        elements: zipLongest(lhs.elements, rhs.elements, default: .zero).map(+))
+    }
+
+    @inlinable
+    public static func - (
+      lhs: Array.TangentVector,
+      rhs: Array.TangentVector
+    ) -> Array.TangentVector {
+      return Array.TangentVector(
+        elements: zipLongest(lhs.elements, rhs.elements, default: .zero).map(-))
+    }
+
+    @inlinable
+    public func moved(
+      along direction: Array.TangentVector
+    ) -> [Element.TangentVector] {
+      precondition(self.elements.count >= direction.elements.count)
```

Comment:
Nit: Our local convention is to omit `self.` when possible. Same for other occurrences.
```suggestion
      precondition(elements.count >= direction.elements.count)
```